src/prepared_statement.c

```diff
@@ -0,0 +1,408 @@
+	HASH_ADD_STR(client->client_prepared_statements, stmt_name, client_ps);
+
+	if (found) {
+		/* Such query was already prepared */
+		HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+		if (server_ps)
+		{
+			/* Statement was already prepared on this server, do not forward packet */
+			slog_debug(client, "handle_parse_command: mapping statement '%s' to '%s' (query '%s')",
+					   client_ps->stmt_name, ps->stmt_name, ps->query_with_params);
+
+			buf = create_parse_complete_packet();
+			res = pktbuf_send_immediate(buf, client);
+			pktbuf_free(buf);
+			return res;
+		}
+	}
+	/* Statement was not prepared on this server, sent modified P packet */
+	slog_debug(client, "handle_parse_command: creating mapping for statement '%s' to '%s' (query '%s')",
+			   client_ps->stmt_name, ps->stmt_name, ps->query_with_params);
+
+	buf = create_parse_packet(ps->stmt_name, &pp);
+
+	/* update stats */
+	client->pool->stats.ps_server_parse_count++;
+
+	/* Track Parse command sent to server */
+	opp = calloc(sizeof *opp, 1);
+	opp->ignore = false;
+	list_append(&server->server_outstanding_parse_packets, &opp->node);
+
+	/* Register statement on server */
+	server_ps = create_server_prepared_statement(ps);
+	if (!register_prepared_statement(server, server_ps))
+		return false;
+
+	return pktbuf_send_queued(buf, server);
+}
+
+bool handle_bind_command(PgSocket *client, PktHdr *pkt)
+{
+	PgSocket *server = client->link;
+	PgBindPacket bp;
+	PgParsePacket pp;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgServerPreparedStatement *server_ps = NULL;
+	PgPreparedStatement* ps;
+	PktBuf *buf;
+	struct OutstandingParsePacket *opp = NULL;
+	int diff;
+	uint32_t len;
+
+	Assert(server);
+
+	if (!unmarshall_bind_packet(client, pkt, &bp))
+		return false;
+
+	/* update stats */
+	client->pool->stats.ps_bind_count++;
+
+	HASH_FIND_STR(client->client_prepared_statements, bp.name, client_ps);
+	if (!client_ps) {
+		slog_error(client,"prepared statement '%s' not found", bp.name);
+		send_pooler_error(client, false, false, "prepared statement not found");
+		return false;
+	}
+	ps = client_ps->ps;
+
+	HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+	if (!server_ps) {
+		/* Statement is not prepared on this link, sent P packet first */
+		slog_debug(server, "handle_bind_command: prepared statement '%s' (query '%s') not available on server, preparing '%s' before bind",
+				   bp.name, ps->query_with_params, ps->stmt_name);
+
+		extract_parse_packet(ps, &pp);
+		buf = create_parse_packet(ps->stmt_name, &pp);
+
+		/* update stats */
+		client->pool->stats.ps_server_parse_count++;
+
+		/* Track Parse command sent to server */
+		opp = calloc(sizeof *opp, 1);
+		opp->ignore = true;
+		list_append(&server->server_outstanding_parse_packets, &opp->node);
+		if (!pktbuf_send_queued(buf, server)) {
+			return false;
+		}
+
+		/* Register statement on server link */
+		server_ps = create_server_prepared_statement(ps);
```

Comment:
CI is reporting 2 memory leaks, both about not freeing the server prepared statement:
```
==13483== VALGRIND-ERROR-BEGIN
==13483== 648 (72 direct, 576 indirect) bytes in 1 blocks are definitely lost in loss record 10 of 11
==13483==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==13483==    by 0x136F17: create_server_prepared_statement (prepared_statement.c:57)
==13483==    by 0x13A4EF: handle_parse_command (prepared_statement.c:191)
==13483==    by 0x11F2A3: handle_client_work (client.c:1061)
==13483==    by 0x11F7ED: client_proto (client.c:1172)
==13483==    by 0x13F44B: sbuf_call_proto (sbuf.c:392)
==13483==    by 0x13FF6F: sbuf_process_pending (sbuf.c:592)
==13483==    by 0x140643: sbuf_main_loop (sbuf.c:772)
==13483==    by 0x1403A9: sbuf_recv_cb (sbuf.c:681)
==13483==    by 0x486F13E: ??? (in /usr/lib/x86_64-linux-gnu/libevent-2.1.so.7.0.0)
==13483==    by 0x486F87E: event_base_loop (in /usr/lib/x86_64-linux-gnu/libevent-2.1.so.7.0.0)
==13483==    by 0x129CF1: main_loop_once (main.c:777)
==13483== 
==13483== VALGRIND-ERROR-END
==13483== VALGRIND-ERROR-BEGIN
==13483== 648 (72 direct, 576 indirect) bytes in 1 blocks are definitely lost in loss record 11 of 11
==13483==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==13483==    by 0x136F17: create_server_prepared_statement (prepared_statement.c:57)
==13483==    by 0x13B531: handle_bind_command (prepared_statement.c:248)
==13483==    by 0x11F302: handle_client_work (client.c:1073)
==13483==    by 0x11F7ED: client_proto (client.c:1172)
==13483==    by 0x13F44B: sbuf_call_proto (sbuf.c:392)
==13483==    by 0x13FF6F: sbuf_process_pending (sbuf.c:592)
==13483==    by 0x140643: sbuf_main_loop (sbuf.c:772)
==13483==    by 0x13E909: sbuf_continue (sbuf.c:242)
==13483==    by 0x12C606: activate_client (objects.c:698)
==13483==    by 0x12CA98: reuse_on_release (objects.c:827)
==13483==    by 0x12D09A: release_server (objects.c:959)
==13483== 
==13483== VALGRIND-ERROR-END
```

Reply:
I believe that `ps_client_free/ps_server_free` should be called at `SL/CL_JUSTFREE` and not at t `SL/CL_FREE`